File: svc-data/src/main/java/com/beancounter/marketdata/portfolio/PortfolioService.java
package com.beancounter.marketdata.portfolio;

import com.beancounter.common.contracts.PortfolioInput;
import com.beancounter.common.exception.BusinessException;
import com.beancounter.common.model.Portfolio;
import com.beancounter.common.model.SystemUser;
import com.beancounter.marketdata.currency.CurrencyService;
import com.beancounter.marketdata.registration.SystemUserService;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@Service
public class PortfolioService {
  private PortfolioRepository portfolioRepository;
  private SystemUserService systemUserService;
  private PortfolioInputAdapter portfolioInputAdapter;

  PortfolioService(
      CurrencyService currencyService,
      PortfolioInputAdapter portfolioInputAdapter,
      PortfolioRepository portfolioRepository,
      SystemUserService systemUserService
  ) {
    this.portfolioRepository = portfolioRepository;
    this.systemUserService = systemUserService;
    this.portfolioInputAdapter = portfolioInputAdapter;
  }

  public Collection<Portfolio> save(Collection<PortfolioInput> portfolios) {
    SystemUser owner = getOrThrow();
    Collection<Portfolio> results = new ArrayList<>();
    portfolioRepository.saveAll(
        portfolioInputAdapter.prepare(owner, portfolios)).forEach(results::add);
    return results;
  }


  private void verifyOwner(SystemUser owner) {
    if (owner == null) {
      throw new BusinessException("Unable to identify the owner");
    }
    if (!owner.getActive()) {
      throw new BusinessException("User is not active");
    }
  }

  private SystemUser getOrThrow() {
    SystemUser systemUser = systemUserService.getActiveUser();
    verifyOwner(systemUser);
    return systemUser;

  }

  private boolean canView(SystemUser systemUser, Portfolio found) {
    return found.getOwner().getId().equals(systemUser.getId());
  }

  public Collection<Portfolio> getPortfolios() {
    SystemUser systemUser = getOrThrow();
    Collection<Portfolio> results = new ArrayList<>();
    Iterable<Portfolio> portfolios = portfolioRepository.findByOwner(systemUser);
    for (Portfolio portfolio : portfolios) {
      results.add(portfolio);
    }
    return results;
  }

  public Portfolio find(String id) {
    SystemUser systemUser = getOrThrow();
    Optional<Portfolio> found = portfolioRepository.findById(id);
    Portfolio portfolio = found.orElseThrow(()
        -> new BusinessException(String.format("Could not find a portfolio with ID %s", id)));

    if (canView(systemUser, portfolio)) {
      return portfolio;
    }

    throw new BusinessException(String.format("Could not find a portfolio with ID %s", id));

  }

  public Portfolio findByCode(String code) {
    SystemUser systemUser = getOrThrow();
    log.debug("Searching on behalf of {}", systemUser.getId());
    Optional<Portfolio> found = portfolioRepository
        .findByCodeAndOwner(code.toUpperCase(), systemUser);
    Portfolio portfolio = found.orElseThrow(()
        -> new BusinessException(
        String.format("Could not find a portfolio with code %s for %s",
            code,
            systemUser.getId())));

    if (canView(systemUser, portfolio)) {
      return portfolio;
    }
    throw new BusinessException(String.format("Could not find a portfolio with code %s", code));

  }

  public Portfolio update(String id, PortfolioInput portfolioInput) {
    Portfolio existing = find(id);
    portfolioInputAdapter.fromInput(portfolioInput, existing);
    return portfolioRepository.save(existing);
  }

}

File: svc-data/src/main/java/com/beancounter/marketdata/portfolio/PortfolioService.kt
package com.beancounter.marketdata.portfolio

import com.beancounter.auth.server.AuthConstants
import com.beancounter.auth.server.AuthConstants.m2mSystemUser
import com.beancounter.common.contracts.PortfoliosResponse
import com.beancounter.common.exception.BusinessException
import com.beancounter.common.exception.ForbiddenException
import com.beancounter.common.input.PortfolioInput
import com.beancounter.common.model.Portfolio
import com.beancounter.common.model.SystemUser
import com.beancounter.common.utils.DateUtils
import com.beancounter.marketdata.registration.SystemUserService
import com.beancounter.marketdata.trn.TrnRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.function.Consumer
import javax.transaction.Transactional

@Service
class PortfolioService internal constructor(
    private val portfolioInputAdapter: PortfolioInputAdapter,
    private val portfolioRepository: PortfolioRepository,
    private val trnRepository: TrnRepository,
    private val systemUserService: SystemUserService,
    private val dateUtils: DateUtils
) {

    fun save(portfolios: Collection<PortfolioInput>): Collection<Portfolio> {
        val owner = orThrow
        val results: MutableCollection<Portfolio> = ArrayList()
        portfolioRepository.saveAll(
            portfolioInputAdapter.prepare(owner, portfolios)
        ).forEach(Consumer { e: Portfolio -> results.add(e) })
        return results
    }

    private fun verifyOwner(owner: SystemUser?) {
        if (owner == null) {
            throw ForbiddenException("Unable to identify the owner")
        }
        if (owner.id == AuthConstants.OAUTH_M2M) {
            return
        }
        if (!owner.active) {
            throw BusinessException("User is not active")
        }
    }

    private val orThrow: SystemUser
        get() {
            if (systemUserService.isServiceAccount()) {
                return m2mSystemUser
            }

            val systemUser = systemUserService.activeUser
            verifyOwner(systemUser)
            return systemUser!!
        }

    fun canView(found: Portfolio): Boolean {
        val systemUser = orThrow
        return systemUser.id == AuthConstants.OAUTH_M2M || found.owner!!.id == systemUser.id
    }

    val portfolios: Collection<Portfolio>
        get() {
            val systemUser = orThrow
            val results: MutableCollection<Portfolio> = ArrayList()
            val portfolios = portfolioRepository.findByOwner(systemUser)
            for (portfolio in portfolios) {
                results.add(portfolio)
            }
            return results
        }

    /**
     * Confirms if the requested portfolio is known. Service side call.
     *
     * @param id pk
     * @return exists or not
     */
    fun verify(id: String): Boolean {
        val found = portfolioRepository.findById(id)
        return found.isPresent
    }

    fun find(id: String): Portfolio {
        val found = portfolioRepository.findById(id)
        val portfolio = found.orElseThrow { BusinessException(String.format("Could not find a portfolio with ID %s", id)) }
        if (canView(portfolio)) {
            return portfolio
        }
        throw BusinessException(String.format("Could not find a portfolio with ID %s", id))
    }

    fun findByCode(code: String): Portfolio {
        val systemUser = orThrow
        log.trace("Searching on behalf of {}", systemUser.id)
        val found = portfolioRepository
            .findByCodeAndOwner(code.toUpperCase(), systemUser)
        val portfolio = found.orElseThrow {
            BusinessException(
                String.format(
                    "Could not find a portfolio with code %s owned by %s",
                    code,
                    systemUser.id
                )
            )
        }
        if (canView(portfolio)) {
            return portfolio
        }
        throw BusinessException(String.format("Could not find a portfolio with code %s", code))
    }

    @Transactional
    fun update(id: String, portfolioInput: PortfolioInput?): Portfolio {
        val existing = find(id)
        return portfolioRepository.save(portfolioInputAdapter.fromInput(portfolioInput!!, existing))
    }

    @Transactional
    fun delete(id: String) {
        val portfolio = find(id)
        trnRepository.deleteByPortfolioId(portfolio.id)
        portfolioRepository.delete(portfolio)
    }

    fun findWhereHeld(assetId: String?, tradeDate: LocalDate?): PortfoliosResponse {
        val recordDate = tradeDate ?: dateUtils.getDate(dateUtils.today())
        val portfolios = portfolioRepository
            .findDistinctPortfolioByAssetIdAndTradeDate(assetId!!, recordDate)
        log.trace("Found {} notional holders for assetId: {}", portfolios.size, assetId)
        return PortfoliosResponse(portfolios)
    }

    companion object {
        private val log = LoggerFactory.getLogger(PortfolioService::class.java)
    }
}
